File: exercicios/projeto-final/skeletons/rickandmorty/react-native/pratica/src/api/rickAndMortyApi.ts
import { CharacterDetail, CharacterSummary } from '../types/character';

const BASE_URL = 'https://rickandmortyapi.com/api';

export async function fetchList(page = 1): Promise<CharacterSummary[]> {
  const res = await fetch(`${BASE_URL}/character?page=${page}`);
  if (!res.ok) throw new Error('Falha ao carregar lista de personagens');
  const body = await res.json();
  return body.results.map((entry: { id: number; name: string }) => ({
    id: entry.id,
    name: entry.name,
  }));
}

export async function fetchDetail(id: number): Promise<CharacterDetail> {
  const res = await fetch(`${BASE_URL}/character/${id}`);
  if (!res.ok) throw new Error(`Falha ao carregar detalhe do personagem ${id}`);
  return res.json();
}

// A própria API filtra por status — sem precisar de endpoint separado de
// "tipo" como o PokeAPI.
export async function fetchNamesByStatus(status: string): Promise<Set<string>> {
  const res = await fetch(`${BASE_URL}/character/?status=${status}`);
  if (!res.ok) throw new Error(`Falha ao carregar status ${status}`);
  const body = await res.json();
  return new Set(body.results.map((entry: { name: string }) => entry.name));
}

File: exercicios/projeto-final/skeletons/rickandmorty/flutter/pratica/lib/api/rickandmorty_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/character.dart';

class RickAndMortyApi {
  static const String baseUrl = 'https://rickandmortyapi.com/api';

  Future<List<CharacterSummary>> fetchList({int page = 1}) async {
    final res = await http.get(Uri.parse('$baseUrl/character?page=$page'));
    if (res.statusCode != 200) {
      throw Exception('Falha ao carregar lista de personagens');
    }
    final body = jsonDecode(res.body) as Map<String, dynamic>;
    final results = body['results'] as List;
    return results
        .map((e) => CharacterSummary.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<CharacterDetail> fetchDetail(int id) async {
    final res = await http.get(Uri.parse('$baseUrl/character/$id'));
    if (res.statusCode != 200) {
      throw Exception('Falha ao carregar detalhe do personagem $id');
    }
    return CharacterDetail.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  /// Retorna o conjunto de nomes de personagens com o status [status]
  /// (alive/dead/unknown) — a própria API filtra, sem precisar de endpoint
  /// separado de "tipo" (diferente do PokeAPI).
  Future<Set<String>> fetchNamesByStatus(String status) async {
    final res = await http.get(Uri.parse('$baseUrl/character/?status=$status'));
    if (res.statusCode != 200) {
      throw Exception('Falha ao carregar status $status');
    }
    final body = jsonDecode(res.body) as Map<String, dynamic>;
    final results = body['results'] as List;
    return results.map((e) => e['name'] as String).toSet();
  }
}
